### ISSUE TITLE
Adding New Models

### ISSUE THREAD

**Issue by heisenbuug** [mlpack/examples#54]
Hey, I am thinking that we can add more models in mlpack model zoo. In GSoC 2019 it was stated that ResNet will be added but I can see that it has not been implemented yet. I was thinking of implementing [SegNet](https://arxiv.org/pdf/1511.00561.pdf) 
We can start by adding basic models for start. I want to help mlpack model zoo grow under GSoC 2020 and I want to start working on it. I  think we can start by adding one model from each category.

- Image Classification

- Object Detection & Image Segmentation

- Body, Face & Gesture Analysis

- Image Manipulation

SegNet will come under Image Segmentation. I would like some help regarding the model selection(which model should be implemented) and discussion regarding that model.

**Comment by zoq** [comment]
Would be great to extend the implemented models, personally I would focus on one area first instead of trying to cover everything, unless somebody likes to implement a specific model. Anyway, maybe it makes sense to start with something less complex like MobileNet or Darknet. Let me know what you think.

**Comment by heisenbuug** [comment]
I agree with you @zoq But I have have doubt. 
MobileNet is a CNN used to classify images whereas Darknet is a NN framework, right? So are we thinking about implementing models or frameworks? 
I think we can start with models as implementing a framework will require more discussion. What do you think?

**Comment by heisenbuug** [comment]
@zoq Hey, is there any upsampling2D algorithm implemented in mlpack? Coz I need to upsample the images in SegNet. Also how should I test the result, I am kinda done with basic architecture of the model but need to test it on some images.

**Comment by zoq** [comment]
> MobileNet is a CNN used to classify images whereas Darknet is a NN framework, right? So are we thinking about implementing models or frameworks?

Darknet is a framework and a model, for example YOLOv2 uses the DarkNet 19 architecture, YOLOv3 DarkNet 53, some of the models configs are here: https://github.com/pjreddie/darknet/tree/master/cfg.

> @zoq Hey, is there any upsampling2D algorithm implemented in mlpack? Coz I need to upsample the images in SegNet. Also how should I test the result, I am kinda done with basic architecture of the model but need to test it on some images.

In this case you would have to implement an `upsampling2D` layer, about testing what about you create a simple test set, that shows that the model works as expected.

**Comment by kartikdutt18** [comment]
@heisenbuug , Can I help out by implementing upsampling layer?

**Comment by heisenbuug** [comment]
> @heisenbuug , Can I help out by implementing upsampling layer?
> I think I have figured out what needs to be done, once it is complete, you can test the model.

If you have already figured out how to implement it then go ahead, till then I will work on completing the model.

> > MobileNet is a CNN used to classify images whereas Darknet is a NN framework, right? So are we thinking about implementing models or frameworks?
> 
> Darknet is a framework and a model, for example YOLOv2 uses the DarkNet 19 architecture, YOLOv3 DarkNet 53, some of the models configs are here: https://github.com/pjreddie/darknet/tree/master/cfg.
> 
> > @zoq Hey, is there any upsampling2D algorithm implemented in mlpack? Coz I need to upsample the images in SegNet. Also how should I test the result, I am kinda done with basic architecture of the model but need to test it on some images.
> 
> In this case you would have to implement an `upsampling2D` layer, about testing what about you create a simple test set, that shows that the model works as expected.

I will look more into Darknet framework and will try to come up with a plan to implement it. I think it will be a long process implementing a framework.

**Comment by kartikdutt18** [comment]
Great, Taking it up then.

**Comment by heisenbuug** [comment]
> Great, Taking it up then.

Hey @kartikdutt18 If you want I think you can take help from here(https://github.com/alexgkendall/caffe-segnet/tree/segnet-cleaned/src/caffe/layers)

Atleast this may help you to understand the flow.

**Comment by heisenbuug** [comment]
@zoq Hey, I found pre-trained model of SegNet [here](url)https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Example_Models/segnet_model_zoo.md 
I saw on mlpack documentation that we can input wieghts in .xml format. I can download weights from [here](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Example_Models/segnet_model_zoo.md) but they are not in xml format. Can we convert these files to .xml 
How should I proceed?

**Comment by zoq** [comment]
> I will look more into Darknet framework and will try to come up with a plan to implement it. I think it will be a long process implementing a framework.

Just to be clear I don't want to implement the framework, I was just saying Darknet is not necessary a framework but also a specific model.

**Comment by zoq** [comment]
> @zoq Hey, I found pre-trained model of SegNet [here](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Example_Models/segnet_model_zoo.md)
> I saw on mlpack documentation that we can input wieghts in .xml format. I can download weights from [here](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Example_Models/segnet_model_zoo.md) but they are not in xml format. Can we convert these files to .xml
> How should I proceed?

@sreenikSS worked on onnx here: https://github.com/sreenikSS/mlpack-Tensorflow-Translator, but as long as not every layer is supported you can't exchange weights between frameworks, that applies to every framework out there.

**Comment by mrityunjay-tripathi** [comment]
I can work on Inception-v3 if no one is already working on it. :)

**Comment by kartikdutt18** [comment]
I think there is a PR open on mlpack repo to implement inception layers.

**Comment by mrityunjay-tripathi** [comment]
> I think there is a PR open on mlpack repo to implement inception layers.

Hmmm... got it. Will see what I can do.

**Comment by heisenbuug** [comment]
> > I will look more into Darknet framework and will try to come up with a plan to implement it. I think it will be a long process implementing a framework.
> 
> Just to be clear I don't want to implement the framework, I was just saying Darknet is not necessary a framework but also a specific model.

Yea @zoq , I got that, even at first I was also looking to implement YOLO, but I ended up on SegNet.


> > I think there is a PR open on mlpack repo to implement inception layers.
> 
> Hmmm... got it. Will see what I can do.

Hey @mrityunjay-tripathi it would be better if you first go through the architecture and see if all the required layers and functions are available with mlpack, if not working on those implemntations would be better. Implementing these layers for whole library and not a specific model would be better so that they can be used to implement other models as well.

**Comment by heisenbuug** [comment]
> > @zoq Hey, I found pre-trained model of SegNet [here](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Example_Models/segnet_model_zoo.md)
> > I saw on mlpack documentation that we can input wieghts in .xml format. I can download weights from [here](https://github.com/alexgkendall/SegNet-Tutorial/blob/master/Example_Models/segnet_model_zoo.md) but they are not in xml format. Can we convert these files to .xml
> > How should I proceed?
> 
> @sreenikSS worked on onnx here: https://github.com/sreenikSS/mlpack-Tensorflow-Translator, but as long as not every layer is supported you can't exchange weights between frameworks, that applies to every framework out there.

@zoq 
I think we need to add more basic functionality to the library so that we can use weights from other pre trained models. Also some layers need to be implemented in order to support implementation of more models. I think by implementing model and creating functionalities for whole lib rather than making them model specific would be a good plan for now.

**Comment by mrityunjay-tripathi** [comment]
> Hey @mrityunjay-tripathi it would be better if you first go through the architecture and see if all the required layers and functions are available with mlpack, if not working on those implemntations would be better. Implementing these layers for whole library and not a specific model would be better so that they can be used to implement other models as well.

Agree. We are missing some of quite basic functionalities.

**Comment by heisenbuug** [comment]
> > Hey @mrityunjay-tripathi it would be better if you first go through the architecture and see if all the required layers and functions are available with mlpack, if not working on those implemntations would be better. Implementing these layers for whole library and not a specific model would be better so that they can be used to implement other models as well.
> 
> Agree. We are missing some of quite basic functionalities.

I am discussing same with other, regarding the implementation of more basic functionalities. Once I am done working on SegNet I will start working on that as well.

**Comment by kartikdutt18** [comment]
Hmm, I think most functionalities are present in mlpack, few layers that are needed already have a PR open and hopefully will be made available soon. I am also working on some layers. It might be a better idea to implement models that you are sure you can implement with the layers that are present in mlpack. As @zoq, suggested it is better to start with small models.

**Comment by zoq** [comment]
> I am discussing same with other, regarding the implementation of more basic functionalities. Once I am done working on SegNet I will start working on that as well.

If there is anything that you think is missing, please open an issue; the machine learning area evolves so fast, there is always something missing, that is required by some models.

**Comment by heisenbuug** [comment]
Hey @zoq I wanted to have a discussion regarding GSoC proposal. Where should I start one? I also mailed all the mentors regarding my thoughts about project...

**Comment by zoq** [comment]
We usually discuss projects on the public mailinglist.

**Comment by heisenbuug** [comment]
@zoq I have mailed on public mailing list, sorry for not following the proper channel here.

**Comment by prince776** [comment]
> Hey, I am thinking that we can add more models in mlpack model zoo. In GSoC 2019 it was stated that ResNet will be added but I can see that it has not been implemented yet. I was thinking of implementing [SegNet](https://arxiv.org/pdf/1511.00561.pdf)

Hi @heisenbuug  Residual module is actually added in `sequential.hpp` in mlpack/methods/ann/layer. I also started working on Residual block but discovered it already existed.

**Comment by heisenbuug** [comment]
> 
> 
> > Hey, I am thinking that we can add more models in mlpack model zoo. In GSoC 2019 it was stated that ResNet will be added but I can see that it has not been implemented yet. I was thinking of implementing [SegNet](https://arxiv.org/pdf/1511.00561.pdf)
> 
> Hi @heisenbuug Residual module is actually added in `sequential.hpp` in mlpack/methods/ann/layer. I also started working on Residual block but discovered it already existed.

Hey @prince776 yea, I saw it, we can discuss where to put these models later, I don't think that will an issue. Also documentation will be done once more models will be added.

**Comment by heisenbuug** [comment]
> 
> 
> Great, Taking it up then.

Hey @kartikdutt18 are you done with this? To complete the model I need that layer and also further testing will be need once I am done with the model...

**Comment by kartikdutt18** [comment]
Hi @heisenbuug, Sorry it's taking me a little time. I have completed Forward pass for that layer but I have my exams in this week so I am only able to work on any issue only during the breaks that I take. I hope to get it done by Sunday (of this week). I hope that's okay with you. Sorry again for causing the delay.

**Comment by heisenbuug** [comment]
Hey Kartik,
Its fine, no worries. Even I am having my exams this week. So I will start
working on it from 12th of this month.

Regards.
Gopi M Tatiraju
